### PR TITLE
fix(sandpack-react): prevent crash upon deleting last visible file

### PR DESCRIPTION
**Pull request by AaronPowell96**
## What kind of change does this pull request introduce?

Alternative to https://github.com/codesandbox/sandpack/pull/659

TLDR: When deleteFile is called, if there are no visibleFiles then open the entryFile. Also set the activeFile is the currently activeFile is deleted.

This change prevents sandpack from erroring out in the scenario where there a no "active files" in view, which results in "code" being undefined among other things.

## What is the current behavior?

image

Currently if you delete the last remaining active file, the whole page crashes.

## What is the new behavior?

New behaviour, the page doesn't crash and you receive a relevant error inside of the sandbox. This seems correct. Seen below (Deleted app.ts from the react.ts template). 
This automatically opens the entry file and presents an error related to user code.

image

## What steps did you take to test this? This is required before we can merge, make sure to test the flow you've updated.

1. Create a local sandpack
2. Use the deleteFile from context on the active file
3. Boom

## Checklist

- [ N/A ] Documentation;
- [ N/A ] Storybook (if applicable);
- [ N/A ] Tests;
- [X] Ready to be merged;

### CODE REVIEW

**Comment by AaronPowell96** [comment]
@danilowoz Do I need to do anything here or is this change reliant on a team member at the moment? Cheers! Not sure if it's waiting for me to do something

**Comment by danilowoz** [comment]
Hey, I'll take a look at this by tomorrow, but in general it looks good to me

Thanks 😄

File: sandpack-react/src/contexts/sandpackContext.tsx [hunk 1]
```diff
@@ -519,12 +519,31 @@ export class SandpackProviderClass extends React.PureComponent<
...
       const newFiles = { ...files };
       delete newFiles[path];
 
+      const remainingVisibleFiles = visibleFiles.filter(
+        (openPath) => openPath !== path
+      );
+      const deletedLastVisibleFile = remainingVisibleFiles.length === 0;
+
+      if (deletedLastVisibleFile) {
+        const nextFile = Object.keys(files)[Object.keys(files).length - 1];
```
Comment (danilowoz) [inline sandpack-react/src/contexts/sandpackContext.tsx:532]: @AaronPowell96 I made a few changes to make the next available file instead of the entry file. I believe it's the safest way to go because I might need to consider that any file can be deleted, even the entry file.

What do you think?

Comment (AaronPowell96) [reply sandpack-react/src/contexts/sandpackContext.tsx:532]: Okay that's better in my opinion too! Nice work

File: sandpack-react/src/contexts/sandpackContext.test.tsx [hunk 2]
```diff
@@ -54,6 +54,38 @@ describe(SandpackProvider, () => {
...
+
+      const instance = root.getInstance();
+
+      instance.runSandpack();
+      instance.deleteFile("/App.js");
+
+      expect(instance.state.activeFile).toBe("/styles.css");
+    });
+
+    it("deletes the activeFile and set the entry file if there no visibleFile left", () => {
```
Comment (AaronPowell96) [inline sandpack-react/src/contexts/sandpackContext.test.tsx:73]: With your change does this test need to be modified as we don't expect just index.js

File: sandpack-react/src/contexts/sandpackContext.test.tsx [hunk 3]
```diff
@@ -54,6 +54,38 @@ describe(SandpackProvider, () => {
...
+          options={{ activeFile: "/App.js", visibleFiles: [] }}
+        />
+      );
+
+      const instance = root.getInstance();
+
+      instance.runSandpack();
+      instance.deleteFile("/App.js");
+
+      expect(instance.state.activeFile).toBe("/index.js");
```
Comment (danilowoz) [inline sandpack-react/src/contexts/sandpackContext.test.tsx:86]:
```suggestion
      expect(instance.state.activeFile).toBe("/package.json");
```

**Comment by danilowoz** [comment]
Thank you @AaronPowell96
